src/cir.rs

```diff
@@ -0,0 +1,311 @@
+                    CirOperand::Predicate(p) => p.execute(ctx, m),
+                }
+        }
+        CirInstruction::NotIns(not) => {
+            let right_val = match &not.right {
+                CirOperand::Index(index) => execute_helper(cir_instructions, *index, ctx, m),
+                CirOperand::Predicate(p) => p.execute(ctx, m),
+            };
+            !right_val
+        }
+        CirInstruction::Predicate(p) => p.execute(ctx, m),
+    }
+}
+
+impl Execute for CirProgram {
+    fn execute(&self, ctx: &mut Context, m: &mut Match) -> bool {
+        execute_helper(&self.instructions, self.instructions.len() - 1, ctx, m)
+    }
+}
+
+impl FieldCounter for CirOperand {
+    fn add_to_counter(&self, map: &mut HashMap<String, usize>) {
+        if let CirOperand::Predicate(p) = &self {
+            *map.entry(p.lhs.var_name.clone()).or_default() += 1
+        }
+    }
+    fn remove_from_counter(&self, map: &mut HashMap<String, usize>) {
+        if let CirOperand::Predicate(p) = &self {
+            let val = map.get_mut(&p.lhs.var_name).unwrap();
+            *val -= 1;
+
+            if *val == 0 {
+                assert!(map.remove(&p.lhs.var_name).is_some());
+            }
+        }
+    }
+}
+
+impl FieldCounter for CirInstruction {
+    fn add_to_counter(&self, map: &mut HashMap<String, usize>) {
+        match self {
+            CirInstruction::AndIns(and) => {
+                and.left.add_to_counter(map);
+                and.right.add_to_counter(map);
+            }
+            CirInstruction::OrIns(or) => {
+                or.left.add_to_counter(map);
+                or.right.add_to_counter(map);
+            }
+            CirInstruction::NotIns(not) => {
+                not.right.add_to_counter(map);
+            }
+            CirInstruction::Predicate(p) => {
+                *map.entry(p.lhs.var_name.clone()).or_default() += 1;
+            }
+        }
+    }
+    fn remove_from_counter(&self, map: &mut HashMap<String, usize>) {
+        match self {
+            CirInstruction::AndIns(and) => {
+                and.left.remove_from_counter(map);
+                and.right.remove_from_counter(map);
+            }
+            CirInstruction::OrIns(or) => {
+                or.left.remove_from_counter(map);
+                or.right.remove_from_counter(map);
+            }
+            CirInstruction::NotIns(not) => {
+                not.right.remove_from_counter(map);
+            }
+            CirInstruction::Predicate(p) => {
+                let val = map.get_mut(&p.lhs.var_name).unwrap();
+                *val -= 1;
+
+                if *val == 0 {
+                    assert!(map.remove(&p.lhs.var_name).is_some());
+                }
+            }
+        }
+    }
+}
+
+impl FieldCounter for CirProgram {
+    fn add_to_counter(&self, map: &mut HashMap<String, usize>) {
+        for instruction in &self.instructions {
+            instruction.add_to_counter(map);
+        }
```

Comment:
https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.for_each

This would be cleaner.

Reply:
fixed